### PR TITLE
[5.2] Index bindings by clause key to facilitate binding removal

### PR DESCRIPTION
**Pull request by Feijs**
Index bindings by the key for the corresponding query clause, and add functions to 
remove clauses and/or binding.

This will simplify removing global scope(s) with binding(s), and make it more robust

### CODE REVIEW

File: src/Illuminate/Database/Query/Builder.php [hunk 1]
```diff
@@ -1920,22 +1930,70 @@ public function setBindings(array $bindings, $type = 'where')
      *
      * @throws \InvalidArgumentException
      */
-    public function addBinding($value, $type = 'where')
+    public function addBinding($value, $type = 'where', $key = null)
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Query/Builder.php]: you need to update the phpdoc

File: src/Illuminate/Database/Query/Builder.php [hunk 2]
```diff
@@ -1920,22 +1930,70 @@ public function setBindings(array $bindings, $type = 'where')
...
         if (! array_key_exists($type, $this->bindings)) {
             throw new InvalidArgumentException("Invalid binding type: {$type}.");
         }
 
-        if (is_array($value)) {
-            $this->bindings[$type] = array_values(array_merge($this->bindings[$type], $value));
-        } else {
-            $this->bindings[$type][] = $value;
-        }
+        if(is_null($key)) $key = $this->getLastKeyOfType($type);
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Query/Builder.php]: cs all wrong

Comment (GrahamCampbell) [inline src/Illuminate/Database/Query/Builder.php]: you MUST use braces

File: src/Illuminate/Database/Query/Builder.php [hunk 3]
```diff
@@ -1920,22 +1930,70 @@ public function setBindings(array $bindings, $type = 'where')
...
+        $this->bindings[$type][$key][] = $value;
 
         return $this;
     }
 
     /**
+     * Remove a binding from the query
+     *
+     * @param  string  $type
+     * @param  int $key
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Query/Builder.php]: cs

File: src/Illuminate/Database/Query/Builder.php [hunk 4]
```diff
@@ -1920,22 +1930,70 @@ public function setBindings(array $bindings, $type = 'where')
...
+     * @return $this
+     */
+    public function removeBinding($type, $key)
+    {
+        unset($this->bindings[$type][$key]);
+        return $this;
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Query/Builder.php]: cs

File: src/Illuminate/Database/Query/Builder.php [hunk 5]
```diff
@@ -1920,22 +1930,70 @@ public function setBindings(array $bindings, $type = 'where')
...
+    }
+
+    /**
+     * Remove a clause and corresponding binding from the query
+     *
+     * @param  string  $type
+     * @param  int $key
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Query/Builder.php]: cs

File: src/Illuminate/Database/Query/Builder.php [hunk 6]
```diff
@@ -1920,22 +1930,70 @@ public function setBindings(array $bindings, $type = 'where')
...
+     *
+     * @param  string  $type
+     * @return int
+     */
+    public function getLastKeyOfType($type)
+    {
+        $var = ($type == 'select') ? "columns" : "{$type}s";
+        
+        $array = $this->$var;
+        if(is_null($array) || empty($array)) return 0;
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Query/Builder.php]: you MUST use braces

Comment (GrahamCampbell) [inline src/Illuminate/Database/Query/Builder.php]: why not just `(! $array)`

**Comment by GrahamCampbell** [comment]
You broke the build too.

**Comment by Feijs** [comment]
Thanks for the comments @GrahamCampbell 

I've discovered the StyleCI and will fix it accordingly.

**Comment by Feijs** [comment]
Fixed the coding style, but made a couple more mistakes on the way. Shall I bundle the changes in another branch and redo the pull request?

**Comment by taylorotwell** [comment]
Sure, please re-do the PR.

**Pull request closed without merging**